internal/app/synchronizer/synchronizer_service.go

```diff
@@ -16,145 +16,541 @@ package synchronizer
-		if err == io.EOF {
-			break
+		select {
+		case s.synchronizeRegistration <- req:
+			return <-req.resp
+		case <-s.startCycle:
+			go func() {
+				s.runCycle()
+				s.startCycle <- struct{}{}
+			}()
 		}
+	}
+}
+
+func (s *synchronizerService) runCycle() {
+	ctx, cancel := WithCancelCause(context.Background())
+
+	m2c := make(chan mAndM6c)
+	m3c := make(chan *pb.Match)
+	m4c := make(chan *pb.Match)
+	m5c := make(chan *pb.Match)
 
-		if err != nil {
-			return err
+	m1c := newCutoffSender(m2c)
+	// m6c, unlike other channels, is specific to a synchronize call.  There are
+	// multiple values in a given cycle.
+
+	var allM1cSent sync.WaitGroup
+
+	registrations := []*registration{}
+	callingCtx := []context.Context{}
+	closedOnCycleEnd := make(chan struct{})
+
+	go func() {
+		fanInFanOut(m2c, m3c, m5c)
+		// Close response channels after all responses have been sent.
+		for _, r := range registrations {
+			close(r.m6c)
 		}
+	}()
+	go s.wrapEvaluator(ctx, cancel, bufferChannel(m3c), m4c)
+	go func() {
+		s.addMatchesToIgnoreList(ctx, cancel, bufferChannel(m4c), m5c)
+		// Wait for ignore list, but not all matches returned, the next cycle
+		// can start now.
+		close(closedOnCycleEnd)
+	}()
 
-		proposals = append(proposals, req.GetMatch())
+	closeRegistration := time.After(s.registrationInterval())
+Registration:
+	for {
+		select {
+		case req := <-s.synchronizeRegistration:
+			allM1cSent.Add(1)
+			callingCtx = append(callingCtx, req.ctx)
+			r := &registration{
+				m1c:        m1c,
+				m6c:        make(chan *pb.Match),
+				cancelMmfs: make(chan struct{}, 1),
+				cycleCtx:   ctx,
+				allM1cSent: &allM1cSent,
+			}
+			registrations = append(registrations, r)
+			req.resp <- r
+		case <-closeRegistration:
+			break Registration
+		}
 	}
 
-	results, err := s.doEvaluateProposals(stream.Context(), proposals, id)
-	if err != nil {
-		return err
+	go func() {
+		for _, ctx := range callingCtx {
+			<-ctx.Done()
+		}
+		cancel(fmt.Errorf("Canceled because all callers were done."))
+	}()
+
+	go func() {
+		allM1cSent.Wait()
+		m1c.cutoff()
+	}()
+
+	cancelProposalCollection := time.AfterFunc(s.proposalCollectionInterval(), func() {
+		m1c.cutoff()
+		for _, r := range registrations {
+			r.cancelMmfs <- struct{}{}
+		}
+	})
+	<-closedOnCycleEnd
+
+	// Clean up in case it was never needed.
+	cancelProposalCollection.Stop()
+}
+
+type registrationRequest struct {
+	resp chan *registration
+	ctx  context.Context
+}
+
+type registration struct {
+	m1c        *cutoffSender
+	allM1cSent *sync.WaitGroup
+	m6c        chan *pb.Match
+	cancelMmfs chan struct{}
+	cycleCtx   context.Context
+}
+
+///////////////////////////////////////
+///////////////////////////////////////
+
+type mAndM6c struct {
+	m   *pb.Match
+	m6c chan *pb.Match
+}
+
+// For incoming mathces, remembers the channel the match needs to be returned to
+// if it passes the synchronizer.
+func fanInFanOut(m2c <-chan mAndM6c, m3c chan<- *pb.Match, m5c <-chan *pb.Match) {
```

Comment:
To me this is the hardest part to understand, it might worth to elaborate what this function is doing